### PR TITLE
UFAL/share submission by email

### LINKED ISSUE #792: UFAL/Share submission by email is not working for a user
## Problem description
The user cannot `Take` a submission by a link.
**Reason**: The user is not authorized to do that, because the BE endpoint is not receiving the workspaceitem id, so it cannot authorize user to `take` that submission.
**Solution**: Send the workspaceitemid from the FE. That id must be added to the `share url`.

### PR DESCRIPTION
**Pull request by milanmajchrak**
| Phases            | MP | MM  |  MB  | MR |   JM | Total  |
|-----------------|----:|----:|----:|-----:|-----:|-------:|
| ETA                  |  0  |  0  |    0 |     0 |      0 |        0 |
| Developing      |  0  |  0  |    0 |    0 |      0 |         0 |
| Review             |  0  |  0  |    0 |    0 |      0 |         0 |
| Total                |   -  |   -  |   -   |  -    |   -    |         0 |
| ETA est.             |      |      |       |       |         |         0 |
| ETA cust.           |   -  |   -  |   -  |   -   |   -     |        0 |

### CODE REVIEW

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/SubmissionController.java [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        validateWorkspaceItem(wsi, null, shareToken);
+
+        if (!authorizeService.authorizeActionBoolean(context, wsi.getItem(), Constants.READ)) {
+            String errorMessage = "The current user does not have rights to view the WorkflowItem";
+            log.error(errorMessage);
+            throw new AccessDeniedException(errorMessage);
+        }
+
+        // Set the owner of the workspace item to the current user
+        EPerson currentUser = context.getCurrentUser();
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/SubmissionController.java:153]: can currentUser be null?

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/WorkspaceItemRestRepository.java [hunk 2]
```diff
@@ -476,6 +481,32 @@ private void uploadFileFromURL(Context context, HttpServletRequest request, Inte
...
+                String errorMessage = "The workspace item with share token:" + shareToken + " does not exist.";
+                log.error(errorMessage);
+                throw new BadRequestException(errorMessage);
+            }
+            if (!authorizeService.authorizeActionBoolean(context, witems.get(0).getItem(), Constants.READ)) {
+                String errorMessage = "The current user does not have rights to view the WorkflowItem";
+                log.error(errorMessage);
+                throw new AccessDeniedException(errorMessage);
+            }
+            // It must return a Page because the FE cannot parse single result from the search endpoint
```
[the remark this answers is not in the document]
Comment (milanmajchrak) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/WorkspaceItemRestRepository.java:500]: done

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/SubmissionController.java (hunk 1 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/SubmissionController.java:153]: validated

**Review by vidiecan — APPROVED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/SubmissionController.java [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        // Without commit the changes are not persisted into the database
+        context.commit();
+        return wsiRest;
+    }
+
+
+
+    private static String generateShareToken() {
+        // UUID generates a 36-char string with hyphens, so we can strip them to get a 32-char string
+        return UUID.randomUUID().toString().replace("-", "").substring(0, 32);
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/SubmissionController.java:174]: why is - a problem?

File: dspace/config/emails/share_submission [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+To pass your submission to another user give them the following link:
+ ${params[0]}
+
+If you have trouble please contact
+${params[1]} or call us at ${params[2]}
+
+
+${params[3]} Team
+
+_____________________________________
```
Comment (vidiecan) [inline dspace/config/emails/share_submission:17]: can't we use something like
${config.get('dspace.name')} ?

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/SubmissionController.java (hunk 3 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/SubmissionController.java:174]: It’s not a problem at all. The token just looks better without the -, that’s all.

File: dspace/config/emails/share_submission (hunk 4 above)
Comment (milanmajchrak) [reply dspace/config/emails/share_submission:17]: Fixed, thank you.

**Review by vidiecan — APPROVED** [review]

**Pull request merged**